### ISSUE TITLE
Union mounted volume created with `docker volume create` no longer work in 17.04 and above

### ISSUE THREAD

**Issue by beenanner** [moby/moby#34107]
**Description**

After upgrading from 17.03 to 17.06 I noticed that a docker volumes that are created with a union mount of 2 directories no longer work. The creation of the volumes will be successful, but once attached to a container the file system is empty within the container. I've attempted to debug the issue with @cpuguy83 but was unable to get the setup to work in 17.06 after several tests. To help tracked it down I've tested the setup in 17.03, 17.04, 17.05, and 17.06 and it appears 17.04 is where the bug was introduced. This was tested on a stock image of Ubuntu 16.04 LTS.

**Steps to reproduce the issue:**
1. Install 17.06 docker
2. mkdir /dir1
3. touch /dir1/file1.txt
4. mkdir /dir2
5. touch /dir2/file2.txt
6. Run `docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test`
7. Run `docker run -it -v test:/test/ ubuntu:16.04 bash`
8. Run `ls -la /test/` and notice files are not displayed

**Describe the results you received:**
Nothing is shown in the /test/ folder

**Describe the results you expected:**
`ls -la /test/` should display
```
root@7a5fd1efe49b:/# ls -la /test/
total 8
drwxr-xr-x  4 root root 4096 Jul 14 01:46 .
drwxr-xr-x 35 root root 4096 Jul 14 01:46 ..
-rw-r--r--  1 root root    0 Jul 14 01:41 file1.txt
-rw-r--r--  1 root root    0 Jul 14 01:41 file2.txt
```
**Additional information you deem important (e.g. issue happens only occasionally):**
```
root@ubuntu-16-04-docker-17-03:/# docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test2
test2
root@ubuntu-16-04-docker-17-03:/# docker run -it -v test2:/test/ ubuntu:16.04 bash
root@7a5fd1efe49b:/# ls -la /test/
total 8
drwxr-xr-x  4 root root 4096 Jul 14 01:46 .
drwxr-xr-x 35 root root 4096 Jul 14 01:46 ..
-rw-r--r--  1 root root    0 Jul 14 01:41 file1.txt
-rw-r--r--  1 root root    0 Jul 14 01:41 file2.txt
```
vs 
```
root@ubuntu-16-04-docker-17-04:/# docker -v
Docker version 17.04.0-ce, build 4845c56
root@ubuntu-16-04-docker-17-04:/# docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test2
test2
root@ubuntu-16-04-docker-17-04:/# docker run -it -v test2:/test/ ubuntu:16.04 bash
root@5147923263e0:/# ls -la /test/
total 8
drwxr-xr-x  2 root root 4096 Jul 14 01:46 .
drwxr-xr-x 35 root root 4096 Jul 14 01:46 ..
```
```
root@ubuntu-16-04-docker-17-05:/# docker -v
Docker version 17.05.0-ce, build 89658be
root@ubuntu-16-04-docker-17-05:/# docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test2
test2
root@ubuntu-16-04-docker-17-05:/# docker run -it -v test2:/test/ ubuntu:16.04 bash
root@5c554c64b4c1:/# ls -la /test/
total 8
drwxr-xr-x  2 root root 4096 Jul 14 01:46 .
drwxr-xr-x 35 root root 4096 Jul 14 01:46 ..
```
and
```
root@ubuntu-16-04-docker-17-06:/# docker -v
Docker version 17.06.0-ce, build 02c1d87
root@ubuntu-16-04-docker-17-06:/# docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test2
test2
root@ubuntu-16-04-docker-17-06:/# docker run -it -v test2:/test/ ubuntu:16.04 bash
root@c0534179291f:/# ls -la /test/
total 8
drwxr-xr-x  2 root root 4096 Jul 14 01:46 .
drwxr-xr-x 35 root root 4096 Jul 14 01:47 ..
```

This also occurs with using the overlay type instead of aufs
```
root@ubuntu-16-04-docker-17-03:/# docker -v
Docker version 17.03.1-ce, build c6d412e
root@ubuntu-16-04-docker-17-03:/# docker volume create --opt 'o=lowerdir=/dir1:/dir2' --opt device=none  --opt type=overlay --name test
test
root@ubuntu-16-04-docker-17-03:/# docker run -it -v test:/test/ ubuntu:16.04 bash
root@2dbe0da29a69:/# ls -la /test/
total 8
drwxr-xr-x  1 root root 4096 Jul 14 01:41 .
drwxr-xr-x 35 root root 4096 Jul 14 01:42 ..
-rw-r--r--  1 root root    0 Jul 14 01:41 file1.txt
-rw-r--r--  1 root root    0 Jul 14 01:41 file2.txt
```
vs
```
root@ubuntu-16-04-docker-17-06:/# docker volume create --opt 'o=lowerdir=/dir1:/dir2' --opt device=none  --opt type=overlay --name test
test
root@ubuntu-16-04-docker-17-06:/# docker run -it -v test:/test/ ubuntu:16.04 bash
root@298dba85aef7:/# ls -la /test/
total 8
drwxr-xr-x  2 root root 4096 Jul 14 01:43 .
drwxr-xr-x 35 root root 4096 Jul 14 01:43 ..
```
**Output of `docker version`:**
```
root@ubuntu-16-04-docker-17-03:/# docker version
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:14:09 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:14:09 2017
 OS/Arch:      linux/amd64
 ```
vs 

```
root@ubuntu-16-04-docker-17-06:/# docker version
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:23:31 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```
**Output of `docker info`:**
```
root@ubuntu-16-04-docker-17-03:/# docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 17.03.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 9
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.8.0-51-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.613 GiB
Name: ubuntu-16-04-docker-17-03
ID: UBQ2:GQ5K:HKCD:I5OY:KJCE:VKUH:HPL5:PDHF:GAEG:GWU5:INP2:QYHO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```
vs
```
root@ubuntu-16-04-docker-17-06:/# docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 17.06.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 9
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.8.0-56-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.613GiB
Name: ubuntu-16-04-docker-17-06
ID: HFYL:P6KL:MT45:5PKD:VLTX:2LEJ:QSNI:R37Y:D3SX:4YYX:BJTA:VMF6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I was able to reproduce this issue in multiple environments AWS, Google, Private DC.

**Comment by beenanner** [comment]
@cpuguy83 pointed out that I tested in 17.03.1. I upgraded to 17.03.2 and confirmed this issues was indeed introduced in 17.04

Here is the output from 17.03.2 where it still works.
```
root@ubuntu-16-04-docker-17-03:/# docker -v
Docker version 17.03.2-ce, build f5ec1e2
root@ubuntu-16-04-docker-17-03:/# docker volume create --opt 'o=br=/dir1=rw:/dir2=ro o=ubda=reval' --opt device=none --opt type=aufs --name test
test
root@ubuntu-16-04-docker-17-03:/# docker run -it -v test:/test/ ubuntu:16.04 bash
root@b139f1e040ff:/# ls -la /test/
total 8
drwxr-xr-x  4 root root 4096 Jul 14 14:58 .
drwxr-xr-x 35 root root 4096 Jul 14 14:58 ..
-rw-r--r--  1 root root    0 Jul 14 01:41 file1.txt
-rw-r--r--  1 root root    0 Jul 14 01:41 file2.txt
```

**Comment by beenanner** [comment]
Unfortunately this inhibits us from upgrading from 17.03 as one of our main applications depend on this functionality. Let me know if there is any additional information I can provide.

**Comment by cpuguy83** [comment]
I am unable to reproduce this issue, but I am on 17.07-rc1, though I don't believe there would be any fixes in it that would change this behavior.

**Comment by cpuguy83** [comment]
If you can test and confirm if this is still an issue for you on 17.07 that would help to know if 1) it is fixed, and 2) if there is something we need to backport

**Comment by beenanner** [comment]
@cpuguy83 will do, I'll give it a go tonight!

**Comment by cpuguy83** [comment]
Thanks!

**Comment by cpuguy83** [comment]
🤞

**Comment by beenanner** [comment]
@cpuguy83 looks like you were right this appears to be fixed in 17.07.

```
root@ubuntu-16-04-docker-17-07:~# docker -v
Docker version 17.07.0-ce-rc3, build 665d244
```

Results:

```
root@bd947fdd39b3:/# ls -la /test/
total 8
drwxr-xr-x  4 root root 4096 Aug 23 02:09 .
drwxr-xr-x 35 root root 4096 Aug 23 02:09 ..
-rw-r--r--  1 root root    0 Aug 23 02:08 file1.txt
-rw-r--r--  1 root root    0 Aug 23 02:08 file2.txt
```

**Comment by beenanner** [comment]
@cpuguy83  looks like this has been backported to 17.06.1-ce, so we are good. Closing out thanks for helping bounce off ways to test in Slack and here! :-) 🥇
